pkg/accounting/accounting.go

```diff
@@ -524,20 +553,63 @@ func (a *Accounting) getAccountingPeer(peer swarm.Address) *accountingPeer {
 	peerData, ok := a.accountingPeers[peer.String()]
 	if !ok {
 		peerData = &accountingPeer{
-			reservedBalance:       big.NewInt(0),
-			shadowReservedBalance: big.NewInt(0),
-			ghostBalance:          big.NewInt(0),
+			reservedBalance:         big.NewInt(0),
+			shadowReservedBalance:   big.NewInt(0),
+			ghostBalance:            big.NewInt(0),
+			totalDebtRepay:          big.NewInt(0),
+			paymentThreshold:        new(big.Int).Set(a.paymentThreshold),
+			paymentThresholdForPeer: new(big.Int).Set(a.paymentThreshold),
+			disconnectLimit:         new(big.Int).Set(a.disconnectLimit),
+			thresholdGrowAt:         new(big.Int).Set(a.thresholdGrowStep),
 			// initially assume the peer has the same threshold as us
-			paymentThreshold: new(big.Int).Set(a.paymentThreshold),
-			earlyPayment:     new(big.Int).Div(new(big.Int).Mul(a.paymentThreshold, big.NewInt(100-a.earlyPayment)), big.NewInt(100)),
-			connected:        false,
+			earlyPayment: percentOf(100-a.earlyPayment, a.paymentThreshold),
+			connected:    false,
 		}
 		a.accountingPeers[peer.String()] = peerData
 	}
 
 	return peerData
 }
 
+// notifyPaymentThresholdUpgrade is used when cumulative debt settled by peer reaches current milestone,
+// to set the next milestone and increase the payment threshold given by 1 * refreshment rate
+// must be called under accountingPeer lock
+func (a *Accounting) notifyPaymentThresholdUpgrade(peer swarm.Address, accountingPeer *accountingPeer) {
+
+	// get appropriate linear growth limit based on whether the peer is a full node or a light node
+	thresholdGrowChange := new(big.Int).Set(a.thresholdGrowChange)
+	if !accountingPeer.fullNode {
+		thresholdGrowChange.Set(a.lightThresholdGrowChange)
+	}
+
+	// if current milestone already passed linear growth limit, set next milestone exponentially
+	if accountingPeer.thresholdGrowAt.Cmp(thresholdGrowChange) >= 0 {
```

Comment:
In the warmup period, connections have their inital, safe-low payment thresholds. The nodes do self throttle, meaning they don't reserve (queue) more requests towards a peer than the payment threshold would allow. As long as the requests are served / fail efficiently (within reasonable time) this throttling effect is negligible. If the initial thresholds are smaller, timing-out / hanging requests can have a more pronounced effect combined with this self throttling.

Later on, the long standing connections that see significant amount of traffic get these self-throttling limits extended, meaning they can work through larger bursts of requests enabling more throughput

The intention is to consecutively decrease the default configured payment threshold to the minimal accepted threshold which is 10 times less than the current default.
In this case, warmup can be considered the time it takes to get a 10x higher payment threshold than the safe-low.
This is the linear phase of threshold growth, which is after having served and get settled for a total of 1800 seconds worth of refreshment rate.
Without cheques this presumes having at least half an hour intensive traffic through the connection.
If cheques are present, this can be shorter, because the same volume of traffic can be served and settled in less time, as the debts are decreased faster than the time based refreshments, and achieving checkpoints earlier are only limited by the speed of the network combined with the self-throttling of concurrently queued requests.


Throughput expected on a single p2p connection with safe-low initial payment threshold:

Reserve limit / number of seconds average request is served in 
 
where
reserve limit ~ minimal payment threshold accepted = 2 * refresh rate ~ 33+ chunks (depending on PO more for higher PO connections)